crates/zir-codegen/src/lib.rs

```diff
@@ -12,10 +12,128 @@ pub mod testing;
 use std::any::Any;
 use std::borrow::Cow;
 use std::collections::HashMap;
+use std::fmt;
 use std::io::Write;
 use std::path::PathBuf;
 
-use zir::mir::Body;
+use thiserror::Error;
+use zir::mir::{Body, Span};
+use zir::ty::TyKind;
+
+// ============================================================================
+// Error Handling
+// ============================================================================
+
+/// Errors that can occur during code generation.
+///
+/// This enum represents structured errors that can happen during the
+/// compilation process, allowing for proper error handling and reporting
+/// instead of panicking.
+#[derive(Debug, Error)]
+pub enum CodegenError {
+    /// Error related to type layout computation.
+    ///
+    /// This occurs when the codegen cannot determine how to lay out a type
+    /// in memory (e.g., unsupported bit widths, invalid struct layouts).
+    #[error("layout error for type `{ty}`: {message}")]
+    LayoutError {
+        /// String representation of the type that caused the error.
+        ty: String,
+        /// Detailed error message.
+        message: String,
+    },
+
+    /// A type is not supported by the backend.
+    ///
+    /// This occurs when encountering types that the backend cannot handle,
+    /// such as arbitrary-precision integers beyond supported widths.
+    #[error("type not supported: `{ty}`")]
+    TypeNotSupported {
+        /// String representation of the unsupported type.
+        ty: String,
+    },
+
+    /// Internal codegen error.
+    ///
+    /// This represents unexpected internal errors that indicate bugs in the
+    /// code generator rather than user errors.
+    #[error("internal codegen error: {message}")]
+    InternalError {
+        /// Detailed error message.
+        message: String,
+    },
+
+    /// Compilation error tied to a source location.
+    ///
+    /// This represents errors that can be traced back to specific source code,
+    /// allowing for precise error reporting to users.
+    #[error("compile error at {span:?}: {message}")]
+    CompileError {
+        /// Source span where the error occurred.
+        span: Span,
+        /// Error message.
+        message: String,
+    },
+
+    /// Unsupported language feature.
+    ///
+    /// This occurs when the code uses a feature that the backend does not
+    /// yet implement.
+    #[error("unsupported feature: {feature}")]
+    UnsupportedFeature {
+        /// Description of the unsupported feature.
+        feature: String,
+    },
+
+    /// Module-level error from the backend.
+    ///
+    /// This wraps backend-specific module errors (e.g., Cranelift's ModuleError).
+    #[error("module error: {message}")]
+    ModuleError {
+        /// Error message from the backend.
+        message: String,
+    },
+}
```

Comment:
Rename to just `Error`, store a real data instead of `String`, error can have a lifetime to satisfy interned conditions